## Pre-primary schools inspection checklist reliability test by Cronbacha Alpha Python script 

### The following Python script was used for the school inspection pilot test conducted in December 2024. The pilot test evaluated inspection  checklists for pre-primary, primary, and secondary schools across 14 regions and city administrations. The dataset includes schools owned by both  government and non-government entities at all three education levels. 

## Cronbach's alpha is a measure of internal consistency, often used to assess the reliability of a survey, test, or questionnaire. The scale ranges from 0 to 1, with higher values indicating greater reliability. Here's a general interpretation of Cronbach's alpha values along with the scenarios and arguments for each category.
1.  α ≥ 0.90: Excellent
-	Scenario: The school inspection checklist is used for critical evaluations, such as determining accreditation status, compliance with legal regulations, or high-stakes decisions.
-	Argument: An alpha in this range ensures that the checklist items are highly consistent in assessing the intended aspects of the school. However, if the alpha is excessively high (close to 1.0), it may suggest redundancy, with multiple items assessing the same criteria unnecessarily.
2.  0.80 ≤ α < 0.90: Good
-	Scenario: The checklist is used for regular school evaluations, such as annual reviews or quality assessments for improvement purposes, where reliability is important but not critical.
-	Argument: A good level of reliability indicates the checklist is consistent and dependable. It balances precision and efficiency without excessive overlap among items, making it suitable for identifying strengths and areas for improvement.
3.  0.70 ≤ α < 0.80: Acceptable
-	Scenario: The checklist is being used for general monitoring purposes, such as district-level reviews or informal assessments, where minor variability in results is tolerable.
-	Argument: While the reliability is adequate for practical use, there may be some inconsistencies in how the checklist items assess the various aspects of the school. Refinements, such as clarifying criteria or reorganizing items, could improve reliability.
4.  0.60 ≤ α < 0.70: Questionable
-	Scenario: The checklist is in the early stages of implementation or being piloted to explore its effectiveness in evaluating schools.
-	Argument: Moderate internal consistency indicates that the checklist provides some useful insights but may not consistently measure the same constructs. Items may need to be revised, clarified, or aligned to improve reliability before formal use.
5.  0.50 ≤ α < 0.60: Poor
-	Scenario: The checklist is unstructured, covering a wide range of unrelated aspects of school performance or being tested with unclear criteria.
-	Argument: Reliability is low, suggesting that the checklist items do not cohesively assess the same aspects of school quality. Significant revisions are needed, such as grouping items under clearer categories or removing irrelevant items.

6.  α < 0.50: Unacceptable
-	Scenario: The checklist is poorly designed, with items that lack coherence or relevance to the intended evaluation of schools.
-	Argument: The checklist is not reliable and should not be used for school inspections. Extensive revisions are necessary to align items with the intended purpose, ensuring they evaluate meaningful and related criteria.
## Key Notes for School Inspection Checklists:
-	Content Coverage: Items should comprehensively cover all relevant domains (e.g., teaching quality, infrastructure, student safety) while avoiding excessive overlap.
-	Clarity and Specificity: Items should be clearly defined and specific to minimize ambiguity and variability in interpretation.
-	Pilot Testing: Piloting the checklist and refining based on feedback can improve internal consistency and ensure the checklist is both reliable and valid.



In [2]:
import pandas as pd
from pingouin import cronbach_alpha
import numpy as np

In [4]:
import pandas as pd

def cronbach_alpha(data):
    """
    Calculate Cronbach's Alpha for a given DataFrame of items.
    Each column represents an item, and each row represents a respondent.

    Parameters:
        data (pd.DataFrame): A DataFrame where each column is an item.

    Returns:
        float: Cronbach's Alpha value.
    """
    # Ensure the DataFrame has at least 2 columns
    if data.shape[1] < 2:
        raise ValueError("The DataFrame must contain at least two columns (items) to calculate Cronbach's Alpha.")

    # Number of items (columns)
    k = data.shape[1]
    
    # Variance of each item (columns)
    item_variances = data.var(axis=0, ddof=1)
    
    # Total variance of the sum of items
    total_variance = data.sum(axis=1).var(ddof=1)
    
    # Cronbach's Alpha formula
    alpha = (k / (k - 1)) * (1 - (item_variances.sum() / total_variance))
    return alpha


# Main function to read CSV and calculate Cronbach's Alpha
if __name__ == "__main__":
    # Path to your CSV file
    csv_file = "C:/Users/Alamerewu/Desktop/pilot test coordinators feedback/google colab analysis/pre-primary/Community engagement.csv"
    
    try:
        # Read the CSV file into a DataFrame
        df = pd.read_csv(csv_file)
        
        # Check if the DataFrame is empty
        if df.empty:
            raise ValueError("The CSV file is empty. Please provide valid data.")
        
        # Ensure only numeric columns are used for the calculation
        numeric_df = df.select_dtypes(include=['number'])
        
        # Check if there are enough numeric columns
        if numeric_df.shape[1] < 2:
            raise ValueError("The CSV must contain at least two numeric columns (items) to calculate Cronbach's Alpha.")
        
        # Calculate Cronbach's Alpha
        alpha = cronbach_alpha(numeric_df)
        print(f"Cronbach's Alpha: {alpha:.4f}")
    
    except FileNotFoundError:
        print(f"Error: The file '{csv_file}' was not found. Please check the file path.")
    except ValueError as ve:
        print(f"Error: {ve}")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")

Cronbach's Alpha: 0.7923


In [5]:
import pandas as pd

def cronbach_alpha(data):
    """
    Calculate Cronbach's Alpha for a given DataFrame of items.
    Each column represents an item, and each row represents a respondent.

    Parameters:
        data (pd.DataFrame): A DataFrame where each column is an item.

    Returns:
        float: Cronbach's Alpha value.
    """
    # Ensure the DataFrame has at least 2 columns
    if data.shape[1] < 2:
        raise ValueError("The DataFrame must contain at least two columns (items) to calculate Cronbach's Alpha.")

    # Number of items (columns)
    k = data.shape[1]
    
    # Variance of each item (columns)
    item_variances = data.var(axis=0, ddof=1)
    
    # Total variance of the sum of items
    total_variance = data.sum(axis=1).var(ddof=1)
    
    # Cronbach's Alpha formula
    alpha = (k / (k - 1)) * (1 - (item_variances.sum() / total_variance))
    return alpha


# Main function to read CSV and calculate Cronbach's Alpha
if __name__ == "__main__":
    # Path to your CSV file
    csv_file = "C:/Users/Alamerewu/Desktop/pilot test coordinators feedback/google colab analysis/pre-primary/conducive environment.csv"
    
    try:
        # Read the CSV file into a DataFrame
        df = pd.read_csv(csv_file)
        
        # Check if the DataFrame is empty
        if df.empty:
            raise ValueError("The CSV file is empty. Please provide valid data.")
        
        # Ensure only numeric columns are used for the calculation
        numeric_df = df.select_dtypes(include=['number'])
        
        # Check if there are enough numeric columns
        if numeric_df.shape[1] < 2:
            raise ValueError("The CSV must contain at least two numeric columns (items) to calculate Cronbach's Alpha.")
        
        # Calculate Cronbach's Alpha
        alpha = cronbach_alpha(numeric_df)
        print(f"Cronbach's Alpha: {alpha:.4f}")
    
    except FileNotFoundError:
        print(f"Error: The file '{csv_file}' was not found. Please check the file path.")
    except ValueError as ve:
        print(f"Error: {ve}")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")

Cronbach's Alpha: 0.9566


In [6]:
import pandas as pd

def cronbach_alpha(data):
    """
    Calculate Cronbach's Alpha for a given DataFrame of items.
    Each column represents an item, and each row represents a respondent.

    Parameters:
        data (pd.DataFrame): A DataFrame where each column is an item.

    Returns:
        float: Cronbach's Alpha value.
    """
    # Ensure the DataFrame has at least 2 columns
    if data.shape[1] < 2:
        raise ValueError("The DataFrame must contain at least two columns (items) to calculate Cronbach's Alpha.")

    # Number of items (columns)
    k = data.shape[1]
    
    # Variance of each item (columns)
    item_variances = data.var(axis=0, ddof=1)
    
    # Total variance of the sum of items
    total_variance = data.sum(axis=1).var(ddof=1)
    
    # Cronbach's Alpha formula
    alpha = (k / (k - 1)) * (1 - (item_variances.sum() / total_variance))
    return alpha


# Main function to read CSV and calculate Cronbach's Alpha
if __name__ == "__main__":
    # Path to your CSV file
    csv_file = "C:/Users/Alamerewu/Desktop/pilot test coordinators feedback/google colab analysis/pre-primary/school culture.csv"
    
    try:
        # Read the CSV file into a DataFrame
        df = pd.read_csv(csv_file)
        
        # Check if the DataFrame is empty
        if df.empty:
            raise ValueError("The CSV file is empty. Please provide valid data.")
        
        # Ensure only numeric columns are used for the calculation
        numeric_df = df.select_dtypes(include=['number'])
        
        # Check if there are enough numeric columns
        if numeric_df.shape[1] < 2:
            raise ValueError("The CSV must contain at least two numeric columns (items) to calculate Cronbach's Alpha.")
        
        # Calculate Cronbach's Alpha
        alpha = cronbach_alpha(numeric_df)
        print(f"Cronbach's Alpha: {alpha:.4f}")
    
    except FileNotFoundError:
        print(f"Error: The file '{csv_file}' was not found. Please check the file path.")
    except ValueError as ve:
        print(f"Error: {ve}")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")

Cronbach's Alpha: 0.9498


In [10]:
import pandas as pd

def cronbach_alpha(data):
    """
    Calculate Cronbach's Alpha for a given DataFrame of items.
    Each column represents an item, and each row represents a respondent.

    Parameters:
        data (pd.DataFrame): A DataFrame where each column is an item.

    Returns:
        float: Cronbach's Alpha value.
    """
    # Ensure the DataFrame has at least 2 columns
    if data.shape[1] < 2:
        raise ValueError("The DataFrame must contain at least two columns (items) to calculate Cronbach's Alpha.")

    # Number of items (columns)
    k = data.shape[1]
    
    # Variance of each item (columns)
    item_variances = data.var(axis=0, ddof=1)
    
    # Total variance of the sum of items
    total_variance = data.sum(axis=1).var(ddof=1)
    
    # Cronbach's Alpha formula
    alpha = (k / (k - 1)) * (1 - (item_variances.sum() / total_variance))
    return alpha


# Main function to read CSV and calculate Cronbach's Alpha
if __name__ == "__main__":
    # Path to your CSV file
    csv_file = "C:/Users/Alamerewu/Desktop/pilot test coordinators feedback/google colab analysis/pre-primary/school leadership and mgmt.csv"
    
    try:
        # Read the CSV file into a DataFrame
        df = pd.read_csv(csv_file)
        
        # Check if the DataFrame is empty
        if df.empty:
            raise ValueError("The CSV file is empty. Please provide valid data.")
        
        # Ensure only numeric columns are used for the calculation
        numeric_df = df.select_dtypes(include=['number'])
        
        # Check if there are enough numeric columns
        if numeric_df.shape[1] < 2:
            raise ValueError("The CSV must contain at least two numeric columns (items) to calculate Cronbach's Alpha.")
        
        # Calculate Cronbach's Alpha
        alpha = cronbach_alpha(numeric_df)
        print(f"Cronbach's Alpha: {alpha:.4f}")
    
    except FileNotFoundError:
        print(f"Error: The file '{csv_file}' was not found. Please check the file path.")
    except ValueError as ve:
        print(f"Error: {ve}")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")
     

Cronbach's Alpha: 0.9549


In [11]:
import pandas as pd

def cronbach_alpha(data):
    """
    Calculate Cronbach's Alpha for a given DataFrame of items.
    Each column represents an item, and each row represents a respondent.

    Parameters:
        data (pd.DataFrame): A DataFrame where each column is an item.

    Returns:
        float: Cronbach's Alpha value.
    """
    # Ensure the DataFrame has at least 2 columns
    if data.shape[1] < 2:
        raise ValueError("The DataFrame must contain at least two columns (items) to calculate Cronbach's Alpha.")

    # Number of items (columns)
    k = data.shape[1]
    
    # Variance of each item (columns)
    item_variances = data.var(axis=0, ddof=1)
    
    # Total variance of the sum of items
    total_variance = data.sum(axis=1).var(ddof=1)
    
    # Cronbach's Alpha formula
    alpha = (k / (k - 1)) * (1 - (item_variances.sum() / total_variance))
    return alpha


# Main function to read CSV and calculate Cronbach's Alpha
if __name__ == "__main__":
    # Path to your CSV file
    csv_file = "C:/Users/Alamerewu/Desktop/pilot test coordinators feedback/google colab analysis/pre-primary/teaching and learning.csv"
    
    try:
        # Read the CSV file into a DataFrame
        df = pd.read_csv(csv_file)
        
        # Check if the DataFrame is empty
        if df.empty:
            raise ValueError("The CSV file is empty. Please provide valid data.")
        
        # Ensure only numeric columns are used for the calculation
        numeric_df = df.select_dtypes(include=['number'])
        
        # Check if there are enough numeric columns
        if numeric_df.shape[1] < 2:
            raise ValueError("The CSV must contain at least two numeric columns (items) to calculate Cronbach's Alpha.")
        
        # Calculate Cronbach's Alpha
        alpha = cronbach_alpha(numeric_df)
        print(f"Cronbach's Alpha: {alpha:.4f}")
    
    except FileNotFoundError:
        print(f"Error: The file '{csv_file}' was not found. Please check the file path.")
    except ValueError as ve:
        print(f"Error: {ve}")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")
     

Cronbach's Alpha: 0.9413


## Based on the Cronbach's alpha test results for the pre-primary school inspection checklists, the following conclusions can be drawn for each thematic area:
###	Conducive Learning Environment (α = 0.7923)
-	Conclusion: The reliability of the items under this thematic area is acceptable, indicating that the checklist effectively captures key elements of a conducive learning environment. While sufficient for practical use, there may be room for improvement in refining the items to achieve higher consistency.
###	School Leadership and Management (α = 0.9549)
-	Conclusion: The reliability of the items is excellent, demonstrating strong internal consistency. This indicates that the items under this thematic area are highly aligned in assessing aspects of school leadership and management, making it suitable for high-stakes evaluations or in-depth analysis.
###	Teaching and Learning (α = 0.9498)
-	Conclusion: The items under this thematic area exhibit excellent reliability. The high Cronbach's alpha suggests that the checklist provides a consistent and reliable assessment of teaching and learning practices, ensuring credibility in evaluations.
###	School Parents and Community Engagement (α = 0.9590)
-	Conclusion: The thematic area demonstrates excellent reliability. The high alpha indicates that the items cohesively measure the involvement of parents and the community in school activities, providing a dependable assessment for this critical area.
###	School Culture (α = 0.9413)
-	Conclusion: The reliability of the items under this thematic area is excellent, suggesting that the checklist effectively evaluates the elements of school culture with a high degree of consistency.

### Overall Conclusion:
- The Cronbach's alpha results indicate that the school inspection checklist is highly reliable across most thematic areas, with alpha values ranging from acceptable to excellent. Specifically:
-	Teaching and Learning, School Leadership and Management, School Parents and Community Engagement, and School Culture exhibit excellent reliability, making them robust for evaluations and critical decision-making.
-	Conducive Learning Environment and School Administration show acceptable to good reliability, suggesting that the checklist is consistent but may benefit from minor refinements in item design to enhance consistency further.

- The high reliability across all thematic areas supports the use of the checklist for school inspections, ensuring dependable assessments of the various dimensions of school performance.



#### From the above finding once we understand that the internal reliability using thematic is better than the general test.  
##### If you want further information contact 
- Alamerewu_Aklilu  Mob. +251913052618,
- email:- menadagi@gmail.com / alamerewu.aklilu@ethernet.edu.et/ nurihagi@gmail.com